.docs/architecture/adr-001-general-parser-architecture.md

```diff
@@ -0,0 +1,234 @@
+    Height() uint64
+
+    // Hash returns the hash of the block.
+    Hash() string
+
+    // Timestamp returns the timestamp of the block.
+    Timestamp() time.Time
+
+    // Proposer returns the address of the block proposer.
+    Proposer() string
+}
+```
+
+### Enqueuer
+
+The Enqueuer interface orchestrates the process of listening for available blocks and enqueueing them into the queue for parsing by workers.
+
+In the current version, two types of enqueuers are implemented:
+
+* Missing Blocks Enqueuer: This enqueuer queries missing blocks from the given start block to the current latest block and enqueues them into the queue.
+* New Blocks Enqueuer: This enqueuer listens for new blocks from the node and enqueues them into the queue.
+
+```go
+type Enqueuer interface {
+    // ListenAndEnqueueBlocks listens for available blocks and enqueues them into the queue.
+    ListenAndEnqueueBlocks(ctx Context, queue BlockQueue) (err error)
+}
+```
+
+### Block Queue
+
+The BlockQueue interface serves as a communication channel between enqueuers and workers within the Juno architecture.
+
+```go
+type BlockQueue interface {
+    // Enqueue enqueues a block into the queue.
+    Enqueue(block Block)
+    
+    // Listen returns a channel to listen for blocks.
+    Listen() <-chan Block
+}
+```
+
+### Node
+
+The Node interface serves as an abstraction layer for interacting with RPC servers or other clients to obtain block data efficiently.
+
+```go
+type Node interface {
+    // LatestBlock retrieves the latest block from the connected node.
+    LatestBlock() (Block, error)
+
+    // Block retrieves the block by the given height
+    Block(height uint64) (Block, error)
+
+    // SubscribeBlocks returns a channel to subscribe to new blocks from the connected node.
+    SubscribeBlocks() (<-chan Block, error)
+}
+```
+
+### Database
+
+The Database interface encapsulates the necessary operations for interacting with the underlying storage system.
+
+```go
+type Database interface {
+    // SaveBlock saves a block into the database.
+    SaveBlock(block Block) error
+
+    // LatestBlock returns the latest block from the database.
+    LatestBlock() (Block, error)
+}
+```
+
+### Worker
+
+The Worker interface defines the functionality for parsing blocks from the queue within Juno's parser architecture.
+
+```go
+type Worker interface {
+    // Start starts the worker to parse blocks from the queue.
+    Start(ctx Context, queue BlockQueue)
+}
+```
+
+### Scheduler
+
+The Scheduler interface facilitates the execution of periodic operations within Juno's parser architecture.
+
+```go
+type PeriodicOperationsModule interface {
+	// RegisterPeriodicOperations allows to register all the operations that will be run on a periodic basis.
+	// The given scheduler can be used to define the periodicity of each task.
+	// NOTE. This method will only be run ONCE during the module initialization.
+	RegisterPeriodicOperations(scheduler Scheduler) error
```

Comment:
What kind of process did we have to launch on a periodic basis, in the last iteration of Juno?

Reply:
It is actually for Callisto for custom module like [gov module](https://github.com/forbole/callisto/blob/cosmos/v0.47.x/modules/gov/handle_periodic_operations.go#L13), I think It would be better to remove it from Juno interfaces as it is actually related to business logic.

Thanks for pointing it out.